tox.ini

```diff
@@ -1,17 +1,36 @@
 [tox]
-envlist = py27
+envlist = py{35,36,37}
 
-[testenv]
-usedevelop = true
+[base]
 deps =
-    -rrequirements.txt
     pyftpdlib
     parameterized
     pytest-xdist
     pytest-cov
-    miniboa
+    miniboa >= 1.0.8
     biplist
+
+[testenv]
+usedevelop = true
+deps =
+    -rrequirements.txt
+    {[base]deps}
 commands =
     py.test {posargs:--tb=short --cov=linkcheck tests}
 setenv =
     LC_ALL=en_US.utf-8
+
+[testenv:py35]
+# test bs4 without line number support
```

Comment:
I would rather use a separate `py35-oldbs4` for this.  Time will come when we drop support Python 3.5, and I don't know if we want to tie dropping of older beautifulsoup versions with that.

You could use tox's [factors](https://tox.readthedocs.io/en/latest/config.html#factors-and-factor-conditional-settings) to adjust dependencies for *-oldbs4 environments.

Reply:
Also, if we care about this, we should have a

```yaml
matrix:
  include:
    - name: old BeautifulSoup4
      python: "3.7"
      script: tox -e py37-oldbs4
```

or similar to add a Travis job for this as well.

Reply:
> I would rather use a separate `py35-oldbs4` for this. Time will come when we drop support Python 3.5, and I don't know if we want to tie dropping of older beautifulsoup versions with that.

Add a separate py35-oldbs4.

> You could use tox's [factors](https://tox.readthedocs.io/en/latest/config.html#factors-and-factor-conditional-settings) to adjust dependencies for *-oldbs4 environments.

The complication is including requirements.txt. You can't override dependency versions by specifying the same package with a different version in the tox.ini.

Reply:
> Also, if we care about this, we should have a
> 
> ```yaml
> matrix:
>   include:
>     - name: old BeautifulSoup4
>       python: "3.7"
>       script: tox -e py37-oldbs4
> ```
> 
> or similar to add a Travis job for this as well.

The way it is now py35 and py35-oldbs4 get run by Travis as the Python 3.5 test. Would adding a matrix help? e.g. to split them into two separate tests?

Reply:
> The complication is including requirements.txt. You can't override dependency versions by specifying the same package with a different version in the tox.ini.

I see setup.py has install_requires as well. So perhaps OK to not use requirements.txt in tox.ini. (Does seem awkward having to maintain both setup.py and requirements.txt).

Reply:
> The complication is including requirements.txt. You can't override dependency versions by specifying the same package with a different version in the tox.ini.

Oh well, scratch that idea, I guess.

+1 for not using requirements.txt in tox.ini (in a separate PR).  Our users (if we ever get any) will install linkchecker with pip install, which will totally ignore our requirements.txt and go along with install_requires from setup.py.  We should test what our users are getting instead of whatever we happened to pin when we found a working set of versions.